File: sdk-javascript/src/recipient/InboundDecryptor.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { decryptUtf8 } from '../crypto/JweEncryption.js';
import type { LocalKeyProvider } from './LocalKeyProvider.js';

/**
 * Composes {@link LocalKeyProvider} with {@link decryptUtf8}. Sister
 * to Java's `InboundDecryptor`, Python's `InboundDecryptor`, and
 * .NET's `InboundDecryptor`. Cross-SDK parity row 12.
 */
export class InboundDecryptor {
  constructor(private readonly keyProvider: LocalKeyProvider) {
    if (!keyProvider) throw new TypeError('keyProvider is required');
  }

  /** Decrypt a JWE compact serialization back to its UTF-8 plaintext. */
  async decrypt(jweCompact: string): Promise<string> {
    if (typeof jweCompact !== 'string') {
      throw new TypeError('jweCompact must be a string');
    }
    const key = await this.keyProvider.getPrivateKey();
    return decryptUtf8(jweCompact, key);
  }
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Recipient/InboundDecryptor.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using HealthFlow.Hfcx.Sdk.Crypto;

namespace HealthFlow.Hfcx.Sdk.Recipient;

/// <summary>
/// Composes <see cref="ILocalKeyProvider"/> with
/// <see cref="JweEncryption"/>'s decrypt path. Sister to Java's
/// <c>InboundDecryptor</c> and Python's <c>InboundDecryptor</c>.
/// Cross-SDK parity row 12.
/// </summary>
public sealed class InboundDecryptor
{
    private readonly ILocalKeyProvider _keyProvider;

    /// <summary>Construct over the given key provider.</summary>
    public InboundDecryptor(ILocalKeyProvider keyProvider)
    {
        ArgumentNullException.ThrowIfNull(keyProvider);
        _keyProvider = keyProvider;
    }

    /// <summary>Decrypt a JWE compact serialization back to its UTF-8 plaintext.</summary>
    public string Decrypt(string jweCompact)
    {
        ArgumentNullException.ThrowIfNull(jweCompact);
        using var key = _keyProvider.GetPrivateKey();
        return JweEncryption.DecryptUtf8(jweCompact, key);
    }
}
